File: apk-extractor/app/src/main/java/net/sarangnamu/apk_extractor/model/Cfg.java
/*
 * Copyright (C) Hanwha S&C Ltd., 2016. All rights reserved.
 *
 * This software is covered by the license agreement between
 * the end user and Hanwha S&C Ltd., and may be
 * used and copied only in accordance with the terms of the
 * said agreement.
 *
 * Hanwha S&C Ltd., assumes no responsibility or
 * liability for any errors or inaccuracies in this software,
 * or any consequential, incidental or indirect damage arising
 * out of the use of the software.
 */
package net.sarangnamu.apk_extractor.model;

import net.sarangnamu.common.BkCfg;
import net.sarangnamu.common.BkString;
import android.content.Context;

public class Cfg extends BkCfg {
    public static final String PATH         = "/apks/";

    private static final String EMAIL       = "email";
    private static final String USERPATH    = "usrPath";
    private static final String SHOW_OPT    = "showOpt";
    private static final String SORT_BY     = "sortBy";
    
    public static final String SORT_DEFAULT = "0";
    public static final String SORT_ALPHABET_ASC = "1";
    public static final String SORT_ALPHABET_DESC = "2";
    public static final String SORT_FIRST_INSTALL_TIME = "3";
    public static final String SORT_LAST_INSTALL_TIME = "4";

    public static String getDownPath(Context context) {
        String usrPath = getUserPath(context);

        if (usrPath == null) {
//            return BkCfg.sdPath() + PATH;
            return "/sdcard" + PATH;
        }

        return BkString.setLastSlash(usrPath);
    }

    public static String getEmail(Context context) {
        return get(context, EMAIL, null);
    }

    public static void setEmail(Context context, String email) {
        set(context, EMAIL, email);
    }

    public static void setUserPath(Context context, String path) {
        set(context, USERPATH, path);
    }

    public static String getUserPath(Context context) {
        return get(context, USERPATH, null);
    }

    public static String getShowOption(Context context) {
        return get(context, SHOW_OPT, "0");
    }

    public static void setShowOption(Context context, String val) {
        set(context, SHOW_OPT, val);
    }

    public static void setSortBy(Context context, String type) {
        set(context, SORT_BY, type);
    }

    public static String getSortBy(Context context) {
        return get(context, SORT_BY, "0");
    }
}

File: apk-extractor/app/src/main/java/net/sarangnamu/apk_extractor/model/Cfg.kt
package net.sarangnamu.apk_extractor.model

import android.annotation.SuppressLint
import android.content.Context
import android.os.Environment
import android.util.Base64
import net.sarangnamu.apk_extractor.MainApp
import net.sarangnamu.common.addLastSlash

/**
 * Created by <a href="mailto:aucd29@gmail.com">Burke Choi</a> on 2017. 11. 30.. <p/>
 */

class Cfg {
    companion object {
        val PATH = "/apks/";

        val EMAIL    = "email"
        val USERPATH = "usrPath"
        val SHOW_OPT = "showOpt"
        val SORT_BY  = "sortBy"

        val SORT_LAST_INSTALL_TIME  = "0"
        val SORT_FIRST_INSTALL_TIME = "3"
        val SORT_ALPHABET_DESC      = "2"
        val SORT_ALPHABET_ASC       = "1"

        fun downloadPath(): String {
            val path = get(USERPATH)

            return path?.addLastSlash() ?: "/sdcard" + PATH
        }

        fun get(key: String, dval: String? = null): String? {
            return MainApp.context().pref(Preference().apply { read(key, dval) })
        }

        fun set(key: String, value: String) {
            MainApp.context().pref(Preference().apply { write(key, value) })
        }
    }
}

inline fun Context.sdPath() =
        Environment.getExternalStorageDirectory().getAbsolutePath()

@SuppressLint("CommitPrefEdits")
inline fun Context.pref(param: Preference): String? {
    val pref = getSharedPreferences("shared.pref", Context.MODE_PRIVATE)

    if (param.write) {
        // write mode
        val editor = pref.edit().putString(param.key, param.value)
        if (param.async) {
            editor.apply()
        } else {
            editor.commit()
        }
    } else {
        // read mode
        val value = pref.getString(param.key, param.value)
        return value?.let { String(Base64.decode(it, Base64.DEFAULT)) } ?: value
    }

    return null
}

class Preference {
    lateinit var key: String
    var value: String? = null
    var write = false
    var async = false

    /**
     * read shared preference
     */
    fun read(key: String, value: String?) {
        data(key, value)
        write = false
    }

    /**
     * write shared preference
     */
    fun write(key: String, value: String?) {
        data(key, value)
        write = true
    }

    private fun data(key: String, value: String?) {
        this.key = key
        this.value = value?.let { Base64.encodeToString(it.toByteArray(), Base64.DEFAULT) } ?: value
    }
}